File: python/backspace-string-compare.py
class Solution:
    # Time complexity: O(n + m) where n and m are the lengths of s and t
    # Space cmoplexity: O(n + m)
    def backspaceCompare(self, s: str, t: str) -> bool:
        ss, tt = [], []
        
        for c in s:
            if c == "#" and ss:
                ss.pop()
            elif c != "#":
                ss.append(c)
        
        for c in t:
            if c == "#" and tt:
                tt.pop()
            elif c != "#":
                tt.append(c)
        print(ss, tt)
        return "".join(ss) == "".join(tt)

File: go/backspace_string_compare.go
package leetcode

// Time complexity: O(n + m) where n and m are the length of s and t
// Space complexity: O(n + m)
func backspaceCompare2(s string, t string) bool {
	build := func(s string) []byte {
		ss := make([]byte, 0, len(s))
		for i := 0; i < len(s); i++ {
			if s[i] == '#' && len(ss) > 0 {
				ss = ss[:len(ss)-1]
			} else if s[i] != '#' {
				ss = append(ss, s[i])
			}
		}
		return ss
	}

	ss, tt := build(s), build(t)
	if len(ss) != len(tt) {
		return false
	}

	for i := 0; i < len(ss); i++ {
		if ss[i] != tt[i] {
			return false
		}
	}
	return true

}
